src/router.py

```diff
@@ -0,0 +1,194 @@
+
+    def handle_arp(self, packet, in_interface):
+        '''
+        Handles ARP requests by unpacking them and constructing a respose packet (if needded).
+
+        :param packet: the entire deconstructed packet
+        :returns: a response packet after arp handling
+        '''
+        interface = self.interfaces[in_interface]
+        arp_request = packet.next_layer
+        if (arp_request.get_opcode() == ArpLayer.OP_IS_AT) or (arp_request.get_dest_ip() != interface.ip):
+            return None
+        
+        response_packet = EthernetLayer() / ArpLayer()
+        response_packet.set_dst(packet.get_src())
+        response_packet.set_src(interface.mac)
+        response_packet.set_ether_type(ArpLayer.ETHR_TYPE)
+        
+        arp_header = response_packet.next_layer
+        arp_header.set_opcode(ArpLayer.OP_IS_AT)
+        arp_header.set_src_mac(interface.mac)
+        arp_header.set_src_ip(interface.ip)
+        arp_header.set_dest_mac(arp_request.get_src_mac())
+        arp_header.set_dest_ip(arp_request.get_src_ip())
+
+        self.arp_tbl[arp_request.get_src_ip()] = arp_request.get_src_mac()
+        return response_packet
+
+    def handle_ip(self, packet, interface):
+        '''
+        Handles IP requests by unpacking them and find the new destination
+        of the packet. Constructing a respose packet.
+
+        :param packet: the entire deconstructed packet
+        :returns: a packet to route
+        '''
+        ip_layer = packet.next_layer
+        dest_ip = socket.inet_ntoa(ip_layer.get_dest_ip())
+        route_interface_ip = self.routing_tbl.route(dest_ip)
+
+        if route_interface_ip is None:
+            # Unreachable. Drop
+            return None
+        
+        elif self.INTERFACE_NAME2IP[interface] == route_interface_ip:
+            if dest_ip not in [inter.ip for inter in self.interfaces.values()]:
+                # Common netwok. Drop
+                return None
+        # ICMP
+        if ip_layer.get_protocol() == ICMPLayer.PROTOCOL_ID:
+            if dest_ip == Router.CLIENT2_IP:
+                packet.set_dst(Router.CLIENT2_MAC)
+                packet.set_src(Router.MAC_ROUTER_NET2)
+            elif dest_ip == Router.CLIENT1_IP:
+                packet.set_dst(Router.CLIENT1_MAC)
+                packet.set_src(Router.MAC_ROUTER_NET1)
+        else:   # UDP/TCP - will be implemented 
+            return None
+
+        return packet
+
+    def observe_traffic(self):
+        '''
+        Gets incoming packets from net1 net2 interfaces and handles them.
+        '''
+        with socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.ntohs(Router.GGP)) as raw_socket:
+            while True:
+                packet, interface = raw_socket.recvfrom(Router.BUF_SIZE)
+                requested_interface = interface[0]
+                self.interfaces.keys()
+                if requested_interface not in list(self.interfaces.keys()):
```

Comment:
Redundant `list` and `.keys` are redundant.

Reply:
list so heavy

list saves what iterator doesnt, python move to iterator writing
in on dict iterates keys already